core/commonMain/src/kotlinx/serialization/encoding/ReorderingCompositeEncoder.kt

```diff
@@ -0,0 +1,240 @@
+
+        private fun bufferEncoding(encoder: Encoder.() -> Unit) {
+            bufferEncoding(descriptor, elementIndex) {
+                encodeInlineElement(descriptor, elementIndex).apply {
+                    if (nullable) {
+                        encodeNotNullMark()
+                    }
+                    encoder()
+                }
+            }
+        }
+
+        override fun encodeNotNullMark() {
+            nullable = true
+        }
+
+        override fun <T : Any> encodeNullableSerializableValue(serializer: SerializationStrategy<T>, value: T?) {
+            bufferEncoding { encodeNullableSerializableValue(serializer, value) }
+        }
+
+        override fun <T> encodeSerializableValue(serializer: SerializationStrategy<T>, value: T) {
+            bufferEncoding { encodeSerializableValue(serializer, value) }
+        }
+
+        override fun encodeBoolean(value: Boolean) {
+            bufferEncoding { encodeBoolean(value) }
+        }
+
+        override fun encodeByte(value: Byte) {
+            bufferEncoding { encodeByte(value) }
+        }
+
+        override fun encodeChar(value: Char) {
+            bufferEncoding { encodeChar(value) }
+        }
+
+        override fun encodeDouble(value: Double) {
+            bufferEncoding { encodeDouble(value) }
+        }
+
+        override fun encodeEnum(enumDescriptor: SerialDescriptor, index: Int) {
+            bufferEncoding { encodeEnum(enumDescriptor, index) }
+        }
+
+        override fun encodeFloat(value: Float) {
+            bufferEncoding { encodeFloat(value) }
+        }
+
+        override fun encodeInt(value: Int) {
+            bufferEncoding { encodeInt(value) }
+        }
+
+        override fun encodeLong(value: Long) {
+            bufferEncoding { encodeLong(value) }
+        }
+
+        @ExperimentalSerializationApi
+        override fun encodeNull() {
+            bufferEncoding { encodeNull() }
+        }
+
+        override fun encodeShort(value: Short) {
+            bufferEncoding { encodeShort(value) }
+        }
+
+        override fun encodeString(value: String) {
+            bufferEncoding { encodeString(value) }
+        }
+
+        override fun beginStructure(descriptor: SerialDescriptor): CompositeEncoder {
+            invalidCall("beginStructure")
+        }
+
+        override fun encodeInline(descriptor: SerialDescriptor): Encoder {
+            invalidCall("encodeInline")
+        }
+
+        private fun invalidCall(methodName: String): Nothing {
+            // This method is normally called by encodeSerializableValue or encodeNullableSerializableValue that is buffered, so we should never go here during buffering as it will be delegated to the concrete CompositeEncoder
+            throw UnsupportedOperationException("$methodName should not be called when reordering fields, as it should be done by encodeSerializableValue or encodeNullableSerializableValue")
+        }
```

Comment:
This is incorrect. It is valid to have an inline/value class have a struct member. That would just result in a nested reordering encoder (delay-created with the result of calling `beginStructure`).

Reply:
The beginStructure is called by the serializer from inside the `encodeSerializableValue`, so it should never be called directly AFAIK 🤔

Reply:
What I mean is that you have an encoder on which `beginStructure` is called. In the implementation it will create/call the reordering composite encoder (and return it), and perhaps also to the delegate encoder (to be passed to the reordering encoder).

Reply:
Ok I got it. Is it such an expected standard use case to encode inlined structures like this `encoder.encodeInlineElement(desc, idx).beginStructure(...)` ? Shouldn't it be expected to be like the serializer plugin (`encoder.encodeInlineElement(desc, idx).encodeSerializableValue(...)`) where structures are always encoded through `encodeSerializableValue` or `encodeNullableSerializableValue` to allow interception and support nullability ?